### DISCUSSION TITLE
Build raylib all with extras as dll?

### DISCUSSION THREAD

**Discussion by Axle-Ozz-i-sofT** [raysan5/raylib#2557] [General]
Hi raysan5 and all

Is it possible to build the entire library v4.0.0 including extras as well as raygui v3.x.x to a single .dll and .so?
If so what compile switches are required (I am using the std makefile/32)?
https://github.com/raysan5/raylib/wiki/CMake-Build-Options

I am sure I seen instructions to be able to include ".\extras" (which included an older version of raygui) as a single dll/so but can't find the file/reference again.

Reason:
I want to create the binders for FreeBASIC. As FB cannot natively compile C headers and needing to recompile the lib.a files in the same C compiler and import a separate static lib for every version of FB to match the C compiler is clunky. If I can build raylib + extras + raygui as a single .dll/.so it will allow it to be used across multiple FB versions.

Regards Axle

**Comment by raysan5** [comment]
@Axle-Ozz-i-sofT extras are being removed in the upcoming `raylib 4.2`. Why do you need them? What extra library do you need?

In any case, `raylib/src/Makefile` has options to add `raygui` and `physac` but the path to those files needs to be provided.

**Comment by Axle-Ozz-i-sofT** [comment]
I apologize as I am not familiar with the library as yet and am uncertain what headers are needed. My default was to try and include ALL into a single dll/so so they are all available if needed as I can only use the dynamic library in FreeBASIC. If the easing.h, physac.h and rmem.h are included in raylib.h already then I guess they are not required for anything.
So that would just leave what is in the raylib root (raylib.h + raymath.h + rlgl.h and raygui.h) to compile to a single shared library file.

Learning raylib + raygui: I will be writing 4 books on programming for students. raylib and raygui is my choice to exemplify the essential graphics programming (graphics.h like such as DSL_BGI.h) with some extra examples that raylib makes available, and raygui to offer some examples of using a GUI widget kit. I will cover other areas in basic such as essential algorithms, Cursers, maybe GNUPlot if I can't do it satisfactorily in raylib, and database such as SQLite. The main focus of this 4 books is on using libraries.
As I make most of the examples from my native C grounding and then transcribe the examples to FB and Python it is simpler for me to learn raylib and raygui and create the examples for all 3 languages at the same time.
The C library is up to date (obviously), the Python 3 bindings appear to be up to date but the FB bindings are a bit dated, clunky and lacking in documentation and I need all 3 languages to be the same version for the book examples. So I am happy to create the dll/so files for FB as well as the bindings and test it all while I am learning the library and writing the books. Obviously the C side will work as expected for the most part so I will see straight away if there is any issue with the FB bindings.

**Reply by raysan5** [reply]
`raygui`, `easings`, `physac` and `rmem` are not part of `raylib`, that's the reason they have been moved out. Bindings shouldn't include those libraries in any way and if they decide to do so, it's up to the binding developers to create specific bindings for them; `raylib.h` deoes not expose the functionality of any of those libraries.

**Comment by Axle-Ozz-i-sofT** [comment]
Sorry about breaking this up as I got an urgent call out.

So essentially I just need to build raylib + raygui and any option additional headers as a single shared library. I will then build the FB binders and likely a single BF raylibgui.bi or raylibex.bi to go with the dll/so. Just to keep it all simple for my students and beginners to drop in and get started with any of the examples in my books.

> In any case, raylib/src/Makefile has options to add raygui and physac but the path to those files needs to be provided.
I think previously I just dropped the extra headers into the raylib root (%path%) and compiled the dll but didn't test or check if the API calls where in the dll. So I will take another shot at it and check with dll export to see if I succeeded.

Last, How far away on the V4.2 and will the changes (extras are being removed in the upcoming raylib 4.2) make much difference to the API from 4.0.0? I may hold off on building the FB headers until the new version. If it's just small API changes I can alter the V4.0.0 raylib.bi to add the changes and get a start on the binders straight away.

Thanks
Axle

**Reply by raysan5** [reply]
`raylib 4.2` has not been published yet but here you can see the list of changes: https://github.com/raysan5/raylib/blob/master/CHANGELOG

**Comment by Axle-Ozz-i-sofT** [comment]
Hi raysan

> extras are being removed in the upcoming raylib 4.2. Why do you need them? What extra library do you need?

I am a little confused here. Does this mean that easings.h, physac.h and rmem.h are deprecated, or do you intend to write them into raylib V4.2.x?
I looked through raylib and didn't find any of the functions for the above, so I am currently unsure where these 3 headers belong. Looking at physac.h "2D Physics library for videogames" I would find this somewhat useful. I would expect the the linked lists and easings would also have a useful place.
I am also looking at "gui_textbox_extended.h" by Vlad Adrian (@Demizdor) and Ramon Santamaria (@raysan5) and would like to make use of it.

My goal is to include and compile all headers to a single shared library as a dll or so for Linux so that I can use it with FB. I have only written a handful of dlls and from a single header so I am stumbling in the dark a little at the moment with how to include the extra headers. I'll read through the provided makefile again and also I took some note previously from the NP++ npes_saved.txt and see if I can nut it out. I am expecting to see __declspec(dllexport) and extern "C" etc in the headers to create a dll/so. Maybe it's somewhere in the makefile to add the extra code with the precompiler, or the above std dll constructions is not required, which sounds odd.

Axle

**Reply by raysan5** [reply]
It means those libraries are extra libraries, they are not part of raylib. The main reason is because some of them are outdated and unmaintained. I prefer to avoid including them with raylib, instead they are provided as separate libraries, each one at its own repo.

`gui_textbox_extended` is also outdated and unmaintained, I don't know if still works, probably not. Feel free to update it yourself if you really need it.

**Reply by Axle-Ozz-i-sofT** [reply]
OK, thanks, that was back to my original question. "It means those libraries are extra libraries, they are not part of raylib." aka easings.h, rmem.h are deprecated. Also gui_textbox_extended.h is in .\extras in the current V3.1 release, so I will take it that gui_textbox_extended.h is also deprecated.
.
Did those 3 headers have, or do they still have any value added for raylib? I can leave them out for now and I am happy to look at updating them if they have a genuine value add for raylib. I am keen on just getting the main components of raylib and raygui compiled as a shared object so I can build the FB bindings so I can use/learn raylib in C. FB and Python at the same time, but I need to have the required dll/so before I start. If I am going to go to the effort to create bindings for FB then I would like to add any useful adittions and do it all in one go.
.
For now though I will be happy to just get raylib + raygui + physac compiled to a win.dll and Linux.so
See below reply...

**Comment by Axle-Ozz-i-sofT** [comment]
Hi raysan

I have tried adding the extra headers into the .\src\Makefile and compiling as a dll but I am having issues even with the default Makefile as it cannot find raygui.c or physac.c

> C:\raylib\raylib\src>mingw32-make -f Makefile PLATFORM=PLATFORM_DESKTOP RAYLIB_LIBTYPE=SHARED
echo #define RAYGUI_IMPLEMENTATION > raygui.c

echo #include "../src/extras/raygui.h" >> raygui.c

gcc -c raygui.c -Wall -D_DEFAULT_SOURCE -Wno-missing-braces -Werror=pointer-arith -fno-strict-aliasing -std=c99 -s -O1 -
Werror=implicit-function-declaration -fPIC -DBUILD_LIBTYPE_SHARED -I. -Iexternal/glfw/include -Iexternal/glfw/deps/mingw
 -DPLATFORM_DESKTOP -DRAYGUI_IMPLEMENTATION
cc1.exe: fatal error: raygui.c: No such file or directory
compilation terminated.
make: *** [Makefile:603: raygui.o] Error 1

Looking at:

> # Compile raygui module
# NOTE: raygui header should be distributed with raylib.h
raygui.o : raygui.c
	$(CC) -c $< $(CFLAGS) $(INCLUDE_PATHS) -D$(PLATFORM) -DRAYGUI_IMPLEMENTATION
raygui.c:
	echo #define RAYGUI_IMPLEMENTATION > raygui.c
	echo #include "$(RAYLIB_MODULE_RAYGUI_PATH)/raygui.h" >> raygui.c

I expect that we are converting raygui.h to (>>) raygui.c but I can't where raygui.c is first declared or created.
Any thoughts on what I am missing here?

**Reply by Axle-Ozz-i-sofT** [reply]
Even attempting to build raygui standalone as dll fails for me...

> raygui is intended to be used as a portable single-file header-only library, to be directly integrated into any C/C++ codebase but some users could require a shared/dynamic version of the library, for example, to create bindings:

    Windows (MinGW, GCC)

copy src/raygui.h src/raygui.c
gcc -o src/raygui.dll src/raygui.c -shared -DRAYGUI_IMPLEMENTATION -DBUILD_LIBTYPE_SHARED -static-libgcc -lopengl32 -lgdi32 -lwinmm -Wl,--out-implib,src/librayguidll.a

CMD line output:
> C:\raylib\raylib\src\extras>gcc -o src/raygui.dll src/raygui.c -shared -DRAYGUI_IMPLEMENTATION -DBUILD_LIBTYPE_SHARED -s
tatic-libgcc -lopengl32 -lgdi32 -lwinmm -Wl,--out-implib,src/librayguidll.a
cc1.exe: fatal error: src/raygui.c: No such file or directory
compilation terminated.

Why can it not find raygui.c ?

Axle

**Reply by orcmid** [reply]
There is no `raygui.c` file.  The `raygui.h` file will do what you want.  You must have a file that includes `raysan5/raygui/src/raygui.h` (wherever you access it from) following a `#define RAYGUI_IMPLEMENTATION`.   You could call your file that has that `raygui.c if you like.  Look past the `raygui.h` comments to the initial code.  You will see how `raygui.h` determines what you are building, and the defines being checked for.

The result will be a `raygui.a` (Linux style) or `raygui.obj` (Microsoft style) that can be incorporated in a DLL.  That will take some  further work on defines, calling styles, etc., and with a linking process that builds the DLL from the raygui compile and all the other compiled components you want in the DLL.   The comments and the code suggest that provision is for a Windows DLL, not a Linux shared library.  (The MngW gcc may handle all that, I wouldn't know.)

There are the usual settings and files you will need to establish so that dynamic linking (DLL) works in the environment you want to use it on.

You're trying to swallow a great deal at once.  Be prepared for that and the speedbumps along the way.

**Reply by raysan5** [reply]
@Axle-Ozz-i-sofT did you edited the `Makefile` to properly configure `RAYLIB_MODULE_RAYGUI_PATH` and `RAYLIB_MODULE_PHYSAC_PATH`? If not, that's the problem.

**Reply by Axle-Ozz-i-sofT** [reply]
**@orcmid** I am trying to compile raygui with raylib to create a single dll. It is not practical to use raylib/raygui in FreeBASIC as a static library.
The instructions say:

> copy src/raygui.h src/raygui.c
gcc -o src/raygui.dll src/raygui.c -shared -DRAYGUI_IMPLEMENTATION -DBUILD_LIBTYPE_SHARED -static-libgcc -lopengl32 -lgdi32 -lwinmm -Wl,--out-implib,src/librayguidll.a

Which is one path I was attempting. Unfortunately raygui.c is not being found.

> You're trying to swallow a great deal at once. Be prepared for that and the speedbumps along the way.
That's OK, I don't mind the effort. I am using GCC, I don't use MS tool chains.

**@raysan5**

> did you edited the Makefile to properly configure RAYLIB_MODULE_RAYGUI_PATH and RAYLIB_MODULE_PHYSAC_PATH? If not, that's the problem.
I did check and it **appeared** to be finding the path to .\src\extras but was still using the relative path from the make file "RAYLIB_MODULE_RAYGUI_PATH ?= $(RAYLIB_SRC_PATH)/extras"
I am setting the paths from a batch file that calls the make32-make -f Makefile in \raylib\src (<- the batch file also SETs the compiler paths.)

I am going to try again this morning and recheck(again) the Win10 R/W permissions for the directories and add the full path to the makefile "RAYLIB_MODULE_RAYGUI_PATH ?= C:\raylib\raylib\src\extras" and set some debug trace "$(info ...)" to see what is happening.

I am currently compiling off of the provided raylib GCC Mingw "https://github.com/raysan5/raylib/releases/download/4.0.0/raylib_installer_v4.0.mingw.exe" But can use any recent version of GCC Mingw-w64 on win, or GCC dev tools on Linux.

**Reply by Axle-Ozz-i-sofT** [reply]
@raysan5 Update
I added -Iextras to the include paths "INCLUDE_PATHS = -I. -Iexternal/glfw/include -Iexternal/glfw/deps/mingw **-Iextras**" with no change. Can't find raygui.c

I moved raygui.h and raygui.c into the src directory ".\raylib\src" and the dll compiled (appears to have) as raygui.o is present.
So I don't know why the compiler is finding ".\raylib\src\external", ".\raylib\src\external\glfw" but **not** ".\raylib\src\extras
¯\_(ツ)_/¯

I'll do a dll export on the raylib dll and check if the function exports are present for raygui API and then I will do the same for physac.h/c (Move them up to the source directory).
Could it bee that the makefile (Make32) is using windows folder paths ".\folder1\subfolder" and I should be changing them to escaped paths ".\\folder1\\subfolder" or maybe using Linux paths in the make file "./folder1/subfolder" ?
Axle

P.S. I just tried with correction to RAYLIB_MODULE_RAYGUI_PATH ?= $C:/raylib/raylib/src/extras and RAYLIB_MODULE_RAYGUI_PATH ?= C:/raylib/raylib/src/extras with raygui.c not found.

**Comment by Axle-Ozz-i-sofT** [comment]
Hi raysan5
Apparently my last update was deleted or didn't post :( It did have some useful information.
In short had to **move the raygui.h/c and physac.h/c to the source directory** to get it to find the files and compile...
I also had to use the latest raygui and physac headers from the repo as the files in ./src/extras did not have the dllexport etc in the headers.

Moving the files to the raylib source seams to have got around the issue and the raygui and physac function definitions/exports are available in libraylibdll.a and raylib.dll

So, I can compile raylib, raygui and pyhsac to a dll under Win GCC. **Success** Just have to repeat that on Ubuntu now.
After I do the FB bindings I will be creating examples for other students as I have mentioned before.. In your opinion do you think there is any additional value for offering physac examples? (I don't want to over complicate and will likely stick to some basic 2D shapes, image rendering, movements and collisions to begin with)

Maybe i'll just kick off with raylib and raygui as a dll/so and do the FB binders and have a look at physac later on and see what the new version 4.2 comes up like (At least I know I can get it to compile now with some band-aids lol).

Thanks for the help so far :)
I have to drop back to OpenGL V 2.2 for VirtualBox compatability so I will see how I go.

Axle

**Reply by raysan5** [reply]
Glad to know you managed to get it properly compiled.
